### ISSUE TITLE
Inconsistent location comparison

### ISSUE THREAD

**Issue by kkarbowiak** [VROOM-Project/vroom#935]
This is related to `libvroom`.

If I have a location created using an index
```c++
auto lhs = vroom::Location(1);
```
and a location created using coordinates
```c++
auto rhs = vroom::Location(vroom::Coordinates{10, 11});
```
I would expect the below to pass
```c++
assert(!(lhs == rhs));
```
but this is not always the case. The comparison code checks presence of index only on left hand side object and assumes the right hand side object to have index as well. This reads to reading from an uninitialised member variable, which is undefined behaviour.

If I flip the arguments
```c++
assert(!(rhs == lhs));
```
the bug is bypassed and result is correct.

**Comment by jcoupey** [comment]
Good catch, I guess this went unnoticed because when parsing json directly we enforce either never or always defining custom index values.

The whole index/coord logic (and hence `operator==`) for locations is a bit awkward but this is the best I could come up with at the time to handle all possible scenarios. I add to check back to recall why we even need `Location::operator==`. :sweat_smile:

**Comment by jcoupey** [comment]
Fixed in #936.
